Core/COObject.m

```diff
@@ -1331,6 +1435,96 @@ - (void)commonDidChangeValueForProperty: (NSString *)key
 
 }
 
+/**
+ * Fast path variant of -commonDidChangeValueForProperty: that only checks
+ * the inserted/removed values
+ */
+- (void)commonDidChangeValueForProperty: (NSString *)key
+							  atIndexes: (NSIndexSet *)indexes
+							withObjects: (NSArray *)objects
+						   mutationKind: (ETCollectionMutationKind)mutationKind
+{
+	ETPropertyDescription *propertyDesc = [_entityDescription propertyDescriptionForName: key];
+	id newValue = [self serializableValueForStorageKey: key];
+	id oldValue = [self popOldCoreObjectRelationshipValueForPropertyDescription: propertyDesc];
+	
+	if (propertyDesc == nil)
+	{
+		[NSException raise: NSInvalidArgumentException
+					format: @"Property %@ is not declared in the metamodel %@ for %@",
+		 key, _entityDescription, self];
+	}
+	
+	if (![self isCoreObjectCollection: newValue])
+	{
+		// Call the slow path.
+		
+		// We rely on COUnsafeRetainedMutableArray automatic removal of duplicates,
+		// otherwise we could use the fast path for non-COPrimitiveCollection multivalues.
+		
+		[self commonDidChangeValueForProperty: key];
+		return;
+	}
+	
+	// Fast path:
+	
+	[(id <COPrimitiveCollection>)newValue setMutable: NO];
+	
+	// We must figure out which objects were added, and which were replaced or removed
+	
+	NSArray *replacedOrRemoved;
+	NSArray *addedObjects;
+	
+	if (mutationKind == ETCollectionMutationKindInsertion)
+	{
+		addedObjects = objects;
+		replacedOrRemoved = @[];
+	}
+	else if (mutationKind == ETCollectionMutationKindRemoval)
+	{
+		addedObjects = @[];
+		replacedOrRemoved = objects;
+	}
+	else if (mutationKind == ETCollectionMutationKindReplacement)
+	{
+		// FIXME: How do we know which objects were replaced if it's an unordered collection?
```

Comment:
ETCollectionMutationKindReplacement is valid for both unordered and ordered collections, but there is a gap as you pointed it out. 

KVO  notifications don't tell you directly about which objects were replaced for unordered collections, so I kept a similar gap for ETCollectionMutationKindReplacement and unordered collections. I think this gap exists in KVO notification, because you can compute it from the old vs new collection differences.

So we can probably compute the replaced objects (COObject takes a snapshot of the old collection in -willChangeValueForProperty:). The old collection is 'oldValue' and the new one 'newValue'. Something like 'replacedObjects = [oldValue minusSet: newValue]' should do the trick, unless I'm missing something.